<a href="https://colab.research.google.com/github/bhavya1234/ICTA--Assignments-and-Case-study/blob/main/Bhavya_S_L_(Assignment_on_Classification).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<font color='orange'>Step 1: Read the dataset into the Python environment

In [ ]:
import pandas as pd
data = pd.read_excel('/content/iris.xls')
data.head()

SL   SW   PL   PW Classification
0  5.1  3.5  1.4  0.2    Iris-setosa
1  4.9  3.0  1.4  0.2    Iris-setosa
2  NaN  3.2  1.3  0.2    Iris-setosa
3  4.6  3.1  1.5  0.2    Iris-setosa
4  5.0  3.6  1.4  0.2    Iris-setosa

<font color='orange'>Step 2: Pre-processing steps

In [ ]:
data.isna().sum()

SL                7
SW                6
PL                6
PW                0
Classification    0
dtype: int64

In [ ]:
# Handle missing values
data['SL'].fillna(data['SL'].mean(), inplace=True)
data['SW'].fillna(data['SW'].mean(), inplace=True)
data['PL'].fillna(data['PL'].mean(), inplace=True)
data['Classification'].fillna(data['Classification'].mode()[0], inplace=True)


<font color='orange'>Step 3: Implement different classification algorithms and evaluate their performance

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import confusion_matrix

# Split the data into features (X) and target variable (y)
X = data.drop('Classification', axis=1)
y = data['Classification']

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# 1. Logistic Regression

In [ ]:
# Create and fit the logistic regression model
lr_model = LogisticRegression()
lr_model.fit(X_train, y_train)
# Make predictions on the testing set
lr_predictions = lr_model.predict(X_test)
# Calculate the accuracy score
lr_accuracy = lr_model.score(X_test, y_test)
print("Logistic Regression Accuracy:", lr_accuracy)

Logistic Regression Accuracy: 1.0


# 2. k-Nearest Neighbors (kNN)

In [ ]:
knn_model = KNeighborsClassifier()
knn_model.fit(X_train, y_train)
knn_predictions = knn_model.predict(X_test)
knn_accuracy = knn_model.score(X_test, y_test)
print("k-Nearest Neighbors Accuracy:",knn_accuracy )

k-Nearest Neighbors Accuracy: 0.9666666666666667


# 3. Confusion Matrix

In [ ]:
cm = confusion_matrix(y_test, knn_predictions)
# Print the accuracy and confusion matrix for each model
print("Confusion Matrix:\n", cm)

Confusion Matrix:
 [[10  0  0]
 [ 0  8  1]
 [ 0  0 11]]


In [ ]:
print("Logistic Regression Accuracy:", lr_accuracy)
print("kNN Accuracy:", knn_accuracy)
print("Confusion Matrix:\n", cm)

Logistic Regression Accuracy: 1.0
kNN Accuracy: 0.9666666666666667
Confusion Matrix:
 [[10  0  0]
 [ 0  8  1]
 [ 0  0 11]]


<font color='orange'>To determine the best model for this specific dataset, we generally consider the model with the highest accuracy. In this case, the Logistic Regression model achieved a perfect accuracy of 1.0, indicating that it correctly predicted all instances in the testing set. However, it's important to note that achieving a perfect accuracy score might indicate overfitting, especially when the dataset is relatively small or unbalanced.

<font color='orange'>On the other hand, the kNN model achieved a slightly lower accuracy of 0.9667, but it still performed very well. The confusion matrix shows that the model made a few misclassifications: one instance of the second class was misclassified as the third class. Overall, the kNN model demonstrates a good balance between accuracy and generalization.

<font color='orange'>In this scenario, the Logistic Regression model provides the best accuracy result of 1.0, but it's recommended to consider other factors such as the dataset size, class distribution, interpretability, and computational complexity when selecting the best model.